sql: panic interface conversion: types.T is types.TOidWrapper, not types.TArray

```
E180307 15:00:09.437959 251041 sql/conn_executor.go:512  [n1,client=127.0.0.1:36492,user=root] a SQL panic has occurred while executing "SELECT CAST('string'::INT2VECTOR AS BIT(135)[])": interface conversion: types.T is types.TOidWrapper, not types.TArray
E180307 15:00:09.517521 251041 util/log/crash_reporting.go:182  [n1,client=127.0.0.1:36492,user=root] a panic has occurred!
E180307 15:00:09.592130 251041 util/log/crash_reporting.go:440  [n1,client=127.0.0.1:36492,user=root] Reported as error c5e2b14e380c4a55b5c1e7cc0c9c9553
panic: interface conversion: types.T is types.TOidWrapper, not types.TArray [recovered]
	panic: panic while executing 1 statements: SELECT CAST(_::INT2VECTOR AS BIT(135)[]); caused by interface conversion: types.T is types.TOidWrapper, not types.TArray

goroutine 251041 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc420ded000, 0x21b16c0, 0xc4211d5280, 0x1ce4800, 0xc4206b3900)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:526 +0x365
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func2(0xc420ded000, 0x21b16c0, 0xc4211d5280)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:487 +0x61
panic(0x1ce4800, 0xc4206b3900)
	/usr/local/go/src/runtime/panic.go:505 +0x229
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.isCastDeepValid(0x21be3a0, 0xc4202682c0, 0x21be280, 0xc421449480, 0x21bc240)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_check.go:270 +0x1d2
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*CastExpr).TypeCheck(0xc4206b3740, 0xc420ded3f0, 0x21be520, 0x3077a00, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_check.go:319 +0x28f
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.TypeCheck(0x21b2a80, 0xc4206b3740, 0xc420ded3f0, 0x21be520, 0x3077a00, 0x1, 0x3078000, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/type_check.go:125 +0xaa
github.com/cockroachdb/cockroach/pkg/sql.(*planner).analyzeExpr(0xc420ded3e0, 0x21b16c0, 0xc4206b3840, 0x21b2a80, 0xc4206b3740, 0xc422c822d0, 0x1, 0x1, 0x3078000, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/analyze_expr.go:72 +0x25e
github.com/cockroachdb/cockroach/pkg/sql.(*planner).computeRender(0xc420ded3e0, 0x21b16c0, 0xc4206b3840, 0x21b2a80, 0xc4206b3740, 0x0, 0x0, 0x21be520, 0x3077a00, 0xc422c822d0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/targets.go:47 +0x13a
github.com/cockroachdb/cockroach/pkg/sql.(*planner).computeRenderAllowingStars(0xc420ded3e0, 0x21b16c0, 0xc4206b3840, 0x21b2a80, 0xc4206b3740, 0x0, 0x0, 0x21be520, 0x3077a00, 0xc422c822d0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/targets.go:77 +0x25b
github.com/cockroachdb/cockroach/pkg/sql.(*planner).initTargets(0xc420ded3e0, 0x21b16c0, 0xc4206b3840, 0xc4219f5b80, 0xc420456a20, 0x1, 0x1, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/render.go:396 +0x325
github.com/cockroachdb/cockroach/pkg/sql.(*planner).SelectClause(0xc420ded3e0, 0x21b16c0, 0xc4206b3840, 0xc420186c60, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/render.go:203 +0x2e4
github.com/cockroachdb/cockroach/pkg/sql.(*planner).Select(0xc420ded3e0, 0x21b16c0, 0xc4206b3840, 0xc4206b3780, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/render.go:121 +0x54f
github.com/cockroachdb/cockroach/pkg/sql.(*planner).newPlan(0xc420ded3e0, 0x21b16c0, 0xc4206b3840, 0x21b3e00, 0xc4206b3780, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:669 +0xd0e
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makePlan(0xc420ded3e0, 0x21b16c0, 0xc4206b3840, 0x21b3e00, 0xc4206b3780, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:266 +0x100
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc420ded000, 0x21b16c0, 0xc4206b3840, 0x21b3e00, 0xc4206b3780, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:505 +0xb30
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc420ded000, 0x21b16c0, 0xc4206b3840, 0x21b3e00, 0xc4206b3780, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:342 +0x9c7
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc420ded000, 0x21b16c0, 0xc4206b3840, 0x21b3e00, 0xc4206b3780, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:83 +0x358
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc420ded000, 0x21b16c0, 0xc4211d5280, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:908 +0x1cc8
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc420344620, 0x21b16c0, 0xc4211d5280, 0xc42065b01b, 0x5, 0xc42065b00d, 0x4, 0x0, 0x0, 0x219d080, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:489 +0xf53
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func3(0xc420344620, 0x21b16c0, 0xc4211d5280, 0xc421543880, 0x5400, 0x15000, 0xc421435390, 0xc420e9bea0, 0xc42191c6e0, 0xc420e9be90)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:260 +0xfc
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:259 +0xf14

```

--- a/pkg/sql/logictest/testdata/logic_test/array
+++ b/pkg/sql/logictest/testdata/logic_test/array
@@ -153,6 +153,9 @@ SELECT ARRAY[1,2,3]::TEXT[]
 ----
 {"1","2","3"}
 
+statement error invalid cast: int\[\] -> INT2VECTOR
+SELECT ARRAY[1,2,3]::INT2VECTOR
+
 # array subscript access
 
 query T

--- a/pkg/sql/sem/tree/type_check.go
+++ b/pkg/sql/sem/tree/type_check.go
@@ -266,6 +266,8 @@ func (expr *CaseExpr) TypeCheck(ctx *SemaContext, desired types.T) (TypedExpr, e
 }
 
 func isCastDeepValid(castFrom, castTo types.T) bool {
+	castFrom = types.UnwrapType(castFrom)
+	castTo = types.UnwrapType(castTo)
 	if castTo.FamilyEqual(types.FamArray) && castFrom.FamilyEqual(types.FamArray) {
 		return isCastDeepValid(castFrom.(types.TArray).Typ, castTo.(types.TArray).Typ)
 	}

